clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -353,45 +353,61 @@ else if (strategy == OffsetResetStrategy.LATEST)
             throw new NoOffsetForPartitionException(partition);
 
         log.debug("Resetting offset for partition {} to {} offset.", partition, strategy.name().toLowerCase(Locale.ROOT));
-        long offset = getOffsetsByTimes(Collections.singletonMap(partition, timestamp)).get(partition).offset();
+        long offset = getOffsetsByTimes(Collections.singletonMap(partition, timestamp), Long.MAX_VALUE).get(partition).offset();
 
         // we might lose the assignment while fetching the offset, so check it is still active
         if (subscriptions.isAssigned(partition))
             this.subscriptions.seek(partition, offset);
     }
 
-    public Map<TopicPartition, OffsetAndTimestamp> getOffsetsByTimes(Map<TopicPartition, Long> timestampsToSearch) {
+    public Map<TopicPartition, OffsetAndTimestamp> getOffsetsByTimes(Map<TopicPartition, Long> timestampsToSearch,
+                                                                     long timeout) {
+        long startMs = time.milliseconds();
+        long remaining = timeout;
         while (true) {
             RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> future = sendListOffsetRequests(timestampsToSearch);
-            client.poll(future);
+            client.poll(future, remaining);
 
             if (future.succeeded())
                 return future.value();
 
-            if (!future.isRetriable())
+            if (future.isDone() && !future.isRetriable())
                 throw future.exception();
 
-            if (future.exception() instanceof InvalidMetadataException)
-                client.awaitMetadataUpdate();
-            else
-                time.sleep(retryBackoffMs);
+            long elapsed = time.milliseconds() - startMs;
+            remaining = timeout - elapsed;
+            if (remaining <= 0)
+                throw new TimeoutException("Failed to get offsets by times in " + timeout + " ms");
+
+            if (future.exception() instanceof InvalidMetadataException) {
+                if (!client.awaitMetadataUpdate(remaining))
+                    throw new TimeoutException("Failed to get offsets by times in " + timeout + " ms");
+            } else
+                time.sleep(Math.min(remaining, retryBackoffMs));
+
+            elapsed = time.milliseconds() - startMs;
+            remaining = timeout - elapsed;
+            if (remaining <= 0)
```

Comment:
If we did as I suggested above, then we could make the inverse of this as the loop condition.